aggregator_api/src/lib.rs

```diff
@@ -0,0 +1,897 @@
+                const ENDPOINT_NAME: &str = "get_task_metrics";
+                |conn: Conn| {
+                    async move {
+                        endpoint_handler::<C, _>(
+                            conn,
+                            ENDPOINT_NAME,
+                            |aggregator_api, conn| async move {
+                                aggregator_api.handle_get_task_metrics(conn).await
+                            },
+                        )
+                        .await
+                    }
+                    .instrument(info_span!(ENDPOINT_NAME))
+                }
+            }),
+    )
+}
+
+async fn endpoint_handler<C: Clock, Fut>(
+    mut conn: Conn,
+    endpoint_name: impl Into<String>,
+    handler: impl Fn(Arc<AggregatorApi<C>>, Conn) -> Fut,
+) -> Conn
+where
+    Fut: Future<Output = Conn>,
+{
+    let start_instant = Instant::now();
+    let aggregator_api: Arc<AggregatorApi<C>> =
+        Arc::clone(conn.state().expect("No AggregatorApi in connection state"));
+
+    conn = aggregator_api.check_auth(conn);
+    if !conn.is_halted() {
+        conn = handler(Arc::clone(&aggregator_api), conn).await;
+    }
+
+    let status: u16 = match conn.status() {
+        Some(status) => status.into(),
+        None => {
+            warn!(path = conn.path(), "No status set on connection");
+            return conn;
+        }
+    };
+
+    aggregator_api.response_time_histogram.record(
+        &Context::current(),
+        start_instant.elapsed().as_secs_f64(),
+        &[
+            KeyValue::new("endpoint", endpoint_name.into()),
+            KeyValue::new("status", status.to_string()),
+        ],
+    );
+    conn
+}
+
+struct AggregatorApi<C: Clock> {
+    ds: Datastore<C>,
+    cfg: Config,
+    response_time_histogram: Histogram<f64>,
+}
+
+impl<C: Clock> AggregatorApi<C> {
+    fn check_auth(&self, conn: Conn) -> Conn {
+        if let Some(authorization_value) = conn.headers().get("authorization") {
+            if let Some(received_token) = authorization_value.as_ref().strip_prefix(b"Basic ") {
+                if self.cfg.auth_tokens.iter().any(|key| {
+                    constant_time::verify_slices_are_equal(received_token, key.as_ref()).is_ok()
+                }) {
+                    // Authorization succeeds.
+                    return conn;
+                }
+            }
+        }
+
+        // Authorization fails.
+        conn.with_status(Status::Unauthorized).halt()
+    }
+
+    async fn handle_get_task_ids(&self, conn: Conn) -> Conn {
+        const PAGINATION_TOKEN_KEY: &str = "pagination_token";
+        let lower_bound = querify(conn.querystring())
+            .into_iter()
+            .find(|&(k, _)| k == PAGINATION_TOKEN_KEY)
+            .map(|(_, v)| TaskId::from_str(v))
+            .transpose();
+        let lower_bound = match lower_bound {
+            Ok(task_id_param) => task_id_param,
+            Err(_) => return conn.with_status(Status::BadRequest).halt(),
```

Comment:
We should log the error, and the same goes for several other places in this file. It'd also be nice to put the error's description in the response body, perhaps as an HTTP problem document, but that can be a future enhancement.